lib/plug/logger_json.ex

```diff
@@ -55,44 +80,168 @@ defmodule Plug.LoggerJSON do
+  response headers) after the request has been processed.
+
+  You can also use anonymous functions for simple cases:
+
+        plug Plug.LoggerJSON,
+          log: :debug,
+          should_log_request_fn: &(&1.request_path in ["/api", "/v1"]),
+          should_log_response_fn: &(&1.request_path not in ["/health", "/metrics"])
+
+  Or share common logic between both functions:
+
+        defp should_log_path?(conn, allowed_paths) do
+          conn.request_path in allowed_paths
+        end
+
+        def should_log_request(conn), do: should_log_path?(conn, ["/api", "/v1"])
+        def should_log_response(conn), do: should_log_path?(conn, ["/api", "/v1", "/health"])
   """
 
   alias Plug.Conn
 
-  @behaviour Plug
-
   require Logger
 
   @typedoc """
-  Type for a plug option
+  Type for a plug option.
   """
-  @type opts :: binary | tuple | atom | integer | float | [opts] | %{opts => opts}
+  @type opt ::
+          :log
+          | :duration_unit
+          | :extra_attributes_fn
+          | :should_log_request_fn
+          | :should_log_response_fn
+          | {:log, atom()}
+          | {:duration_unit, :nanoseconds | :microseconds | :milliseconds}
+          | {:extra_attributes_fn, (Plug.Conn.t() -> map())}
+          | {:should_log_request_fn, (Plug.Conn.t() -> boolean())}
+          | {:should_log_response_fn, (Plug.Conn.t() -> boolean())}
 
   @typedoc """
-  Type for time
+  Type for a list of plug options.
   """
-  @type time :: {non_neg_integer(), non_neg_integer(), non_neg_integer()}
+  @type opts :: list(opt)
 
-  @spec init(opts) :: opts
+  @typedoc """
+  Type for a timestamp with `{mega_secs, secs, micro_secs}`.
+  """
+  @type os_timestamp :: {non_neg_integer(), non_neg_integer(), non_neg_integer()}
+
+  @spec init(opts()) :: opts()
   def init(opts), do: opts
 
-  @spec call(Plug.Conn.t(), opts) :: Plug.Conn.t()
+  @spec call(Plug.Conn.t(), opts() | atom()) :: Plug.Conn.t()
   def call(conn, level_or_opts) when is_atom(level_or_opts) do
-    call(conn, level: level_or_opts)
+    call(conn, log: level_or_opts)
   end
 
   def call(conn, opts) do
     level = Keyword.get(opts, :log, :info)
     start = :os.timestamp()
 
+    # Store logging info in conn private for access in exception handlers
+    conn = Conn.put_private(conn, :plug_logger_json_opts, opts)
+    conn = Conn.put_private(conn, :plug_logger_json_start, start)
+    conn = Conn.put_private(conn, :plug_logger_json_level, level)
+
+    # Log request phase if enabled
+    if should_log_request?(conn, opts) do
+      log(conn, level, start, opts)
+    end
+
+    # Register before_send callback for response phase
     Conn.register_before_send(conn, fn conn ->
-      :ok = log(conn, level, start, opts)
+      if should_log_response?(conn, opts) do
+        # Mark that we're in the before_send callback
+        conn = Conn.put_private(conn, :plug_logger_json_before_send, true)
+        log(conn, level, start, opts)
```

Comment:
The `plug_logger_json_before_send` is only needed on the `log/4` call? The `conn` that will be returned from this function won't have that field set.